### PR TITLE
New benchmark measuring QNode _construct efficiency.

### PR DESCRIPTION
**Pull request by smite**
**Description of the Change:**

* Fixes a bug in `benchmark/bm_entangling_layers.py`
* Adds ``benchmark/bm_mutable_complicated_params.py``

  This new benchmark aims to measure the efficiency of `JacobianQNode._construct` and `BaseQNode._construct` for mutable QNodes.

  It uses an extreme case where the QNode has lots of primary parameters but relatively few auxiliary parameters, and only a few of the primary parameters are actually used in the circuit, to make the node construction the execution bottleneck.

**Benefits:**

The benchmark underlines some algorithmic inefficiencies in how the QNodes are constructed and the partial derivative methods chosen in the current `master` branch.

https://github.com/XanaduAI/pennylane/pull/510 provides a modified construction logic which results in a better performance in this benchmark.

This branch (equivalent to `master`)
```
Commit: f6588a1d Merge branch 'master' into bm_mutable_node_complicated_params
Timing: 'mutable qnode, complicated primary params' benchmark on default.qubit
Timing per loop: [0.23343417 0.23362751 0.23217291 0.23272017 0.23161589]
```

#510
```
Commit: 794c1a42 More renaming re: primary parameters for consistency.
Timing: 'mutable qnode, complicated primary params' benchmark on default.qubit
Timing per loop: [0.10682147 0.10648436 0.10559171 0.10535492 0.10574054]
```

#510 with another small optimization that speeds up `JacobianQNode._best_method`
```
Commit: cd862259 More efficient handling of BaseQNode.primary_deps.
Timing: 'mutable qnode, complicated primary params' benchmark on default.qubit
Timing per loop: [0.07194923 0.07283727 0.06603256 0.06573229 0.06618677]
```

### CODE REVIEW

File: benchmark/bm_mutable_complicated_params.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+class Benchmark(bu.BaseBenchmark):
+    """
+    This benchmark attempts to measure the efficiency of :meth:`JacobianQNode._construct` for
+    mutable QNodes, using an extreme case where the QNode has lots of primary parameters with
+    a complicated nested structure, but relatively few auxiliary parameters, and only a few
+    of the primary parameters are actually used in the circuit.
+
+    When the QNode is constructed, a VariableRef is built for each primary parameter,
+    and the qfunc re-evaluated. In this test this meant to be time-consuming, but it is only
```
Comment (co9olguy) [inline benchmark/bm_mutable_complicated_params.py]: seems to be a typo here?

File: benchmark/bm_mutable_complicated_params.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    strictly necessary if the auxiliary parameters change.
+    The main reasons why there are significant differences in the execution speed of this test
+    between different PL commits:
+
+      * :meth:`BaseQNode._construct` should only reconstruct the QNode if the auxiliary params
+        have changed.
+      * Most of the primary params are not used in the circuit, hence
+        :meth:`JacobianQNode._construct` should efficiently figure out that partial derivatives
+        wrt. them are always zero.
```
Comment (co9olguy) [inline benchmark/bm_mutable_complicated_params.py]: very good point :+1:

File: benchmark/bm_mutable_complicated_params.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            """Recursively builds a tree structure with n levels."""
+            if n <= 0:
+                # the leaves are arrays
+                return np.random.randn(2)
+            # the other nodes have two branches and a scalar
+            return [create_params(n - 1), create_params(n - 1), np.random.randn()]
+
+        p = create_params(n)
+
+        def evaluate(aux):
```
Comment (co9olguy) [inline benchmark/bm_mutable_complicated_params.py]: What is `aux` here?

File: benchmark/bm_mutable_complicated_params.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            # check the result
+            assert np.allclose(res, np.cos(p[aux][2]))
+
+        # first evaluation and construction
+        evaluate(0)
+        # evaluate the node several times more with different aux params
+        # (it does not matter if p changes or not, the VariableRefs handle it)
+        for _ in range(1, 10):
+            evaluate(1)  # use the same auxiliary params
+            # evaluate(i % 2)  # contrast: changing the aux params every time negates most possible speedups
```
Comment (co9olguy) [inline benchmark/bm_mutable_complicated_params.py]: Is this line meant to be uncommented? Is it a different case (which maybe we want to have a flag for?)

File: benchmark/bm_mutable_complicated_params.py (hunk 1 above)
Comment (smite) [reply benchmark/bm_mutable_complicated_params.py]: Yeah.

File: benchmark/bm_mutable_complicated_params.py (hunk 3 above)
Comment (smite) [reply benchmark/bm_mutable_complicated_params.py]: Auxiliary parameter values.

File: benchmark/bm_mutable_complicated_params.py (hunk 4 above)
Comment (smite) [reply benchmark/bm_mutable_complicated_params.py]: No, IMO there's little point in using it as a benchmark, it's just a comment that explains how/why the benchmark was designed the way it is.

Comment (smite) [reply benchmark/bm_mutable_complicated_params.py]: I moved it into the other comment.

File: benchmark/bm_entangling_layers.py [hunk 5]
```diff
@@ -56,5 +56,5 @@ def benchmark(self, n=3):
 
         qnode = bu.create_qnode(circuit, self.device, mutable=False)
         qnode(init_weights, features=features)
-        qnode.jacobian(init_weights, {"features": features})
+        qnode.jacobian((init_weights,), {"features": features})
```
Comment (smite) [inline benchmark/bm_entangling_layers.py]: I decided to sneak in a quick bugfix to another benchmark.

**Comment by smite** [comment]
@josh Yeah, I added it to the first post.

File: benchmark/bm_entangling_layers.py (hunk 5 above)
Comment (josh146) [reply benchmark/bm_entangling_layers.py]: @trbromley noted this as well! He also noted the need to mark it `mutable` in order to get it working?

Comment (smite) [reply benchmark/bm_entangling_layers.py]: That's a different thing, a workaround to a bug in the current `master` branch. The bug is fixed by #510.

**Pull request merged**